File: src/se/clau/intellijlux/LuxReference.java
package se.clau.intellijlux;

import com.intellij.codeInsight.lookup.*;
import com.intellij.openapi.project.Project;
import com.intellij.openapi.util.TextRange;
import com.intellij.psi.*;
import se.clau.intellijlux.psi.LuxIdent;
import org.jetbrains.annotations.*;

import java.util.*;

public class LuxReference
        extends PsiReferenceBase<PsiElement>
        implements PsiPolyVariantReference {
  private String key;

  public LuxReference(@NotNull PsiElement element, TextRange textRange) {
    super(element, textRange);
    key = element.getText().substring(textRange.getStartOffset(), textRange.getEndOffset());
  }

  @NotNull
  @Override
  public ResolveResult[] multiResolve(boolean incompleteCode) {
    Project project = myElement.getProject();
    final List<LuxIdent> properties = LuxUtil.findIdentifiers(project);
    List<ResolveResult> results = new ArrayList<>();
    for (LuxIdent property : properties) {
      results.add(new PsiElementResolveResult(property));
    }
    return results.toArray(new ResolveResult[results.size()]);
  }

  @Nullable
  @Override
  public PsiElement resolve() {
    ResolveResult[] resolveResults = multiResolve(false);
    return resolveResults.length == 1 ? resolveResults[0].getElement() : null;
  }

  @NotNull
  @Override
  public Object[] getVariants() {
    Project project = myElement.getProject();
    List<LuxIdent> idents = LuxUtil.findIdentifiers(project);
    List<LookupElement> variants = new ArrayList<>();
    for (final LuxIdent idnt : idents) {
      if (idnt.getText() != null && idnt.getText().length() > 0) {
        variants.add(LookupElementBuilder.create(idnt)
                .withIcon(LuxIcons.FILE)
                .withTypeText(idnt.getContainingFile().getName())
        );
      }
    }
    return variants.toArray();
  }
}

File: src/se/clau/intellijlux/LuxReference.kt
package se.clau.intellijlux

import com.intellij.codeInsight.lookup.LookupElement
import com.intellij.codeInsight.lookup.LookupElementBuilder
import com.intellij.openapi.util.TextRange
import com.intellij.psi.*
import org.jetbrains.annotations.NotNull
import se.clau.intellijlux.LuxUtil.findIdentifiers


class LuxReference(
  @NotNull element: PsiElement,
  textRange: TextRange
) : PsiReferenceBase<PsiElement?>(element, textRange), PsiPolyVariantReference {

  private var name: String? = null

  init {
    name = element.text.substring(textRange.startOffset, textRange.endOffset)
  }

  override fun multiResolve(incompleteCode: Boolean): Array<ResolveResult> {
    val project = myElement!!.project
    val properties = findIdentifiers(project)
    val results: MutableList<ResolveResult> = ArrayList()
    for (property in properties) {
      results.add(PsiElementResolveResult(property))
    }
    return results.toTypedArray()
  }

  override fun resolve(): PsiElement? {
    val resolveResults = multiResolve(false)
    return if (resolveResults.size == 1) resolveResults[0].element else null
  }

  override fun getVariants(): Array<Any> {
    val project = myElement!!.project
    val idents = findIdentifiers(project)
    val variants: MutableList<LookupElement> = ArrayList()

    for (the_ident in idents) {
      if (the_ident.text != null && the_ident.text.isNotEmpty()) {
        variants.add(
          LookupElementBuilder.create(the_ident)
            .withIcon(LuxIcons.FILE)
            .withTypeText(the_ident.containingFile.name)
        )
      }
    }
    return variants.toTypedArray()
  }

}